src/geode/mesh/core/regular_grid.cpp

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (c) 2019 - 2020 Geode-solutions
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#include <geode/mesh/core/regular_grid.h>
+
+#include <geode/basic/attribute_manager.h>
+#include <geode/basic/bitsery_archive.h>
+#include <geode/basic/pimpl_impl.h>
+
+#include <geode/geometry/point.h>
+namespace geode
+{
+    template < index_t dimension >
+    class RegularGrid< dimension >::Impl
+    {
+    public:
+        Impl() = default;
+
+        Impl( Point< dimension > origin,
+            std::array< index_t, dimension > cells_number,
+            std::array< double, dimension > cells_size )
+            : origin_( std::move( origin ) ),
+              cells_number_( std::move( cells_number ) ),
+              cells_size_( std::move( cells_size ) )
+        {
+            index_t nb_cells{ 1 };
+            for( const auto c : cells_number )
+            {
+                nb_cells *= c;
+            }

```

Comment:
It is really a detail point. In this kind of process, I always thinking of `std::accumulate` function. Which version is better using for instance `nb_cells = std::accumulate(cells_number.begin(), cells_number.end(), 1, std::multiplies<index_t>());` or the one you propose? I would say: your version is more clear, and maybe it is exactly the same for the compiler.